### PR TITLE
Fix message_queue job loops and incorrect schedule dates

### PR DESCRIPTION
**Pull request by anton-vlasenko**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | 
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) |
| BC breaks? |
| Deprecations? | Deprecates `MessageQueueModel::addToQueue` and `MessageQueueModel::rescheduleMessage` in favour of new methods. The old methods require you to pass date interval in a custom format and these methods throw an exception if you pass 'H'. Currently we have a major bug because of this ambiguity. We reschedule failed messages 15 months ahead (instead of 15 minutes). The new methods accept \DateInterval object. This removes any ambiguity related to date intervals.

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

1. We are experiencing a loop when processing campaign `message_queue` jobs. Same jobs get injected over and over again which leads to db overload.
2. Also, we get incorrect scheduled_date value when rescheduling failed queue message.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Make sure that the email service provider you are using is working. You have to be able to send emails from your Mautic instance.
2. Go to settings -> Emails, set `Default Frequency Rule` to 1 message per day. 
3. Create 2 separate template emails.
4. Create a segment.
5. Create a campaign based on the segment above.
6. Add 2 "Send email" actions to that campaign. Both should be marketing emails. Use emails you created on step 3.
First email should be sent immediately, another one should be sent with 1 minute delay. Publish the campaign.
7. Add a contact that meets the criteria of your segment. It should be added to the segment and it has to have a valid email address.
8. Check that you've got the first email. The second one should be put in a message queue (message_queue table).
9. Go to settings -> Emails, set Default Frequency Rule to 0 message per day (or remove 0 and just leave the empty field), save the settings.
10. Unpublish the second email template you created on step 3, save it.
11. Wait at least 24 hours.
Or you have an option to modify the scheduled_date table. In this case you don't have to wait 24 hours. To do so go to the `message_queue` table, find the queue entity that was created on the step 8 (it should be the last one in the table). Update the value in the `scheduled_date` column to be less or equal the current datetime. This SQL query can do it for you:
```sql
UPDATE mautic_message_queue 
SET 
    scheduled_date = DATE_SUB(NOW(), INTERVAL 2 DAY)
ORDER BY id DESC LIMIT 1;
```
12. Run 
```bash
php app/console mautic:messages:send
```
to trigger message queue processing.
13. Go to the `message_queue` table. Observe that the `scheduled_date` value was scheduled to 15 months in the future. This is wrong. If the message is failed we don't have to wait so long.

#### Steps to test this PR:
1. Repeat steps 1-12 from "Steps to reproduce the bug".
2. Go to the `message_queue` table. Observe that the `scheduled_date` value was incremented by 15 minutes. If the message fails (because of the unpublished email template), we have to reschedule it early and not in 15 months.

### CODE REVIEW

**Comment by npracht** [comment]
@anton-vlasenko can you have a look at Travis issue please ?

File: app/bundles/ChannelBundle/Model/MessageQueueModel.php
Comment (escopecz) [inline app/bundles/ChannelBundle/Model/MessageQueueModel.php]:
```suggestion
     * Adds messages to the queue.
```
CS check is failing because of this.

Comment (npracht) [reply app/bundles/ChannelBundle/Model/MessageQueueModel.php]: @anton-vlasenko can you please? I don't have rights on the source repo to commit instead of you.

**Comment by escopecz** [comment]
@npracht Anton is my colleague. I believe someone from the community (not Acquia) should review/test this PR before merge. We already reviewed and tested it internally.

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 versions (except for major security purpose) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.0.1 candidate.**

## How to do?
1. Check if the bug still present in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please comment on your PR: “This PR only applies to Mautic 2.x”

**Please report results by commenting on your PR to make us administration easier.**

_In case your bugfix only apply to Mautic 2, we'll consider adding it in an extra Mautic 2 version._
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by npracht** [comment]
Hello @anton-vlasenko could you please test mautic 3.0.0-beta2 and tell me if the issue is still existing?
- If no, i'll tag this issue at `M2 only relevant`
- If yes, could you please rebase your branch against 3.x branch ?

Waiting for your feedback, i would very love to merge it in 3.0.1 if still relevant. Thanks !

**Comment by anton-vlasenko** [comment]
I believe the issue is still exists.
I've resolved the conflict.

File: app/bundles/ChannelBundle/Tests/Model/MessageQueueModelTest.php
Comment (escopecz) [inline app/bundles/ChannelBundle/Tests/Model/MessageQueueModelTest.php]:
```suggestion
class MessageQueueModelTest extends \PHPUnit\Framework\TestCase
```

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
@anton-vlasenko I have applied the changes that @escopecz suggested as it was a small fix.

@mabumusa1 @dennisameling as this is in use on production and covered by tests, I think we should be OK to go with a code review by a non-acquian if one of you has the time.

**Review by mabumusa1 — DISMISSED** [review]
I walked through the whole process to test this PR, it works as expected and it is good to be merged with 3.1.1

**Comment by mabumusa1** [comment]
@RCheesley please fix travis issues, then this PR is RTC

**Comment by mabumusa1** [comment]
@cla-bot check

**Comment by escopecz** [comment]
I fixed the issues reported by Travis

**Pull request merged**
